### PR TITLE
add spec infer related into prometheus metrics.

### PR DESCRIPTION
**Pull request by leiwen83**
And add a new boost_ratio metric used to directly show how much spec infer would help in saving decoding steps.

### CODE REVIEW

**Comment by leiwen83** [comment]
cc @cadedaniel

**Comment by cadedaniel** [comment]
will take a look Monday. btw, how is this different from system efficiency metric? (boost ratio == num_spec_tokens+1 * system efficiency?)

File: vllm/engine/metrics.py [hunk 1]
```diff
@@ -59,7 +59,19 @@ def __init__(self, labelnames: List[str], max_model_len: int):
             name="vllm:cpu_cache_usage_perc",
             documentation="CPU KV-cache usage. 1 means 100 percent usage.",
             labelnames=labelnames)
-
+        #   Speculative infer Status in %
```
Comment (robertgshaw2-neuralmagic) [inline vllm/engine/metrics.py:62]: Please add better descriptions.

Please name these:
- `vllm:spec_decode_system_efficiency`
- `vllm:spec_decode_boost_ratio`
- `vllm:spec_decode_draft_acceptance_rate`

What is the difference between these?

Comment (robertgshaw2-neuralmagic) [reply vllm/engine/metrics.py:62]: We prefer to use Counters >> Gauges

Is there a way these metrics could be expressed as Counters with the `rate` function used in PromQL to compute the rates?

**Review by robertgshaw2-redhat — CHANGES_REQUESTED** [review]
Please update the Grafana Dashboard

**Comment by robertgshaw2-neuralmagic** [comment]
> will take a look Monday. btw, how is this different from system efficiency metric? (boost ratio == num_spec_tokens+1 * system efficiency?)

+1

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks for the contribution! It would be great to have these metrics flowing through prometheus!

**Comment by leiwen83** [comment]
> will take a look Monday. btw, how is this different from system efficiency metric? (boost ratio == num_spec_tokens+1 * system efficiency?)

the new boost_ratio would express more accurate expression at how much system is benefit from spec info, as there is case that spec info give no proposal, like no matching in ngram or seqlen+spec exceed over model length.

Furthermore, with the new dynamic spec coming https://github.com/vllm-project/vllm/issues/4565, the k would not be constant one, so that we may need accumulate actual token emitted comparing with the steps.

File: vllm/engine/metrics.py (hunk 1 above)
Comment (leiwen83) [reply vllm/engine/metrics.py:62]: maybe we could directly express the total emitted token, along with steps number? so that user could do the cal they want with those counters?

**Comment by leiwen83** [comment]
@cadedaniel @robertgshaw2-neuralmagic 
Any comment for the latest PR change? :)

**Comment by cadedaniel** [comment]
asking @LiuXiaoxuanPKU if she has bandwidth to review the PR. the approach looks good to me, concerns are 1) we should make sure the top-level metrics make sense to users (not just to us as developers), 2) the naming of the metrics collection seems weird

**Comment by robertgshaw2-neuralmagic** [comment]
I can review + merge tomorrow

**Comment by leiwen83** [comment]
Hi @robertgshaw2-neuralmagic @cadedaniel ,

How is going with the spec related metric, have we got the conclusion for how to make it happen? ;)
The metric is critical to us as a direct feedback reflecting how well current spec sys is doing.

**Comment by cadedaniel** [comment]
thanks & sorry this slipped. I might have time tomorrow to finish review. cc @LiuXiaoxuanPKU and @comaniac who might have bandwidth.

File: vllm/model_executor/layers/rejection_sampler.py [hunk 2]
```diff
@@ -318,10 +328,18 @@ def _create_output(
...
+        output_with_bonus_tokens = torch.cat(
+            [output_with_bonus_tokens, non_spec_token_ids])
+
         self.num_accepted_tokens += accepted.sum()
         self.num_emitted_tokens += (output_with_bonus_tokens != -1).sum()
         self.num_draft_tokens += batch_size * k
 
+        # batch_size here maybe 0, as there is case
+        # where no proposal is generated
+        self.num_specs += output_with_bonus_tokens.size()[0]
```
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:341]: Is this right? even if there are no specs this can be nonzero

File: vllm/model_executor/layers/rejection_sampler.py [hunk 3]
```diff
@@ -318,10 +328,18 @@ def _create_output(
         output.mul_(~after_false_mask).add_(
             recovered_token_ids.mul(after_false_mask))
 
+        # concat spec and nonspec token
+        output_with_bonus_tokens = torch.cat(
+            [output_with_bonus_tokens, non_spec_token_ids])
```
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:333]: `num_emitted_tokens` will include non-spec tokens, making the metric useless in capturing spec system efficiency (100% efficiency is defined every token accepted plus the bonus token)

File: vllm/spec_decode/spec_decode_worker.py [hunk 4]
```diff
@@ -311,20 +311,20 @@ def _verify_tokens(
...
+            non_spec_token_ids=non_spec_token_ids,
         )
 
-        # Append output tokens from non-speculative sequences to
-        # the accepted token ids tensor.
-        non_spec_token_ids = non_spec_token_ids.expand(-1, max_proposal_len +
-                                                       1).clone()
-        non_spec_token_ids[:, 1:] = -1
-        accepted_token_ids = torch.cat(
-            [accepted_token_ids, non_spec_token_ids])
```
Comment (cadedaniel) [inline vllm/spec_decode/spec_decode_worker.py:327]: btw this cat should stay here; the format required for the cat is determined here `original_indices = spec_indices + non_spec_indices`

File: vllm/spec_decode/metrics.py [hunk 5]
```diff
@@ -162,6 +168,7 @@ def _collect_rejsample_metrics(
 
         return SpecDecodeWorkerMetrics(
             num_spec_tokens=k,
+            num_specs=self._aggregate_num_specs,
```
Comment (cadedaniel) [inline vllm/spec_decode/metrics.py:171]: we can calculate this with `draft_tokens // k`, don't need to record

Comment (cadedaniel) [reply vllm/spec_decode/metrics.py:171]: see `get_max_num_emitted_tokens`

**Review by cadedaniel — COMMENTED** [review]
I have bandwidth to review this now. The Prometheus stuff looks good to me, have concerns on the metric definition and how we collect them.

File: vllm/model_executor/layers/rejection_sampler.py (hunk 2 above)
Comment (leiwen83) [reply vllm/model_executor/layers/rejection_sampler.py:341]: For the ngram case, if we don't match anything, then there would be no proposal.

Comment (cadedaniel) [reply vllm/model_executor/layers/rejection_sampler.py:341]: OK. These metrics should work for draft model as well.

**Comment by leiwen83** [comment]
@cadedaniel 
I submit a rebased PR, which keep the concat logic as before. num_spec is made to aggregate "batch_size*k" number.
